packages/react-query/src/__tests__/prefetch.test.tsx

```diff
@@ -0,0 +1,353 @@
+        </div>
+      )
+    }
+
+    function App() {
+      usePrefetchQuery({ ...queryOpts, queryFn: anotherSpy })
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <Suspended />
+        </React.Suspense>
+      )
+    }
+
+    await queryClient.fetchQuery(queryOpts)
+    const rendered = renderWithClient(queryClient, <App />)
+
+    expect(rendered.queryByText('fetching: true')).not.toBeInTheDocument()
+    expect(
+      rendered.getByText('data: Prefetch is really nice!'),
+    ).toBeInTheDocument()
+    expect(anotherSpy).not.toHaveBeenCalled()
+  })
+
+  it('should not create a suspense waterfall if prefetch is fired', async () => {
+    const firstQueryOpts = {
+      queryKey: queryKey(),
+      queryFn: vi.fn().mockImplementation(async () => {
+        await sleep(10)
+
+        return 'Prefetch is nice!'
+      }),
+    }
+
+    const secondQueryOpts = {
+      queryKey: queryKey(),
+      queryFn: vi.fn().mockImplementation(async () => {
+        await sleep(10)
+
+        return 'Prefetch is really nice!!'
+      }),
+    }
+
+    const thirdQueryOpts = {
+      queryKey: queryKey(),
+      queryFn: vi.fn().mockImplementation(async () => {
+        await sleep(10)
+
+        return 'Prefetch does not create waterfalls!!'
+      }),
+    }
+
+    function FirstSuspended() {
+      const state = useSuspenseQuery(firstQueryOpts)
+
+      return (
+        <div>
+          <div>data: {String(state.data)}</div>
+        </div>
+      )
+    }
+
+    function SecondSuspended() {
+      const state = useSuspenseQuery(secondQueryOpts)
+
+      return (
+        <div>
+          <div>data: {String(state.data)}</div>
+        </div>
+      )
+    }
+
+    function ThirdSuspended() {
+      const state = useSuspenseQuery(thirdQueryOpts)
+
+      return (
+        <div>
+          <div>data: {String(state.data)}</div>
+        </div>
+      )
+    }
+
+    function App() {
+      usePrefetchQuery(firstQueryOpts)
+      usePrefetchQuery(secondQueryOpts)
+      usePrefetchQuery(thirdQueryOpts)
+
+      const [showSecond, setShowSecond] = React.useState(false)
+      const [showThird, setShowThird] = React.useState(false)
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <FirstSuspended />
+          <button onClick={() => setShowSecond(true)}>showSecond</button>
+          {showSecond ? <SecondSuspended /> : null}
+          <button onClick={() => setShowThird(true)}>showThird</button>
+          {showThird ? <ThirdSuspended /> : null}
+        </React.Suspense>
+      )
+    }
+
+    const rendered = renderWithClient(queryClient, <App />)
+
+    await waitFor(() => rendered.getByText('Loading...'))
+    await waitFor(() => rendered.getByText('data: Prefetch is nice!'))
+    fireEvent.click(rendered.getByText('showSecond'))
+    expect(rendered.queryByText('Loading...')).not.toBeInTheDocument()
+    await waitFor(() => rendered.getByText('data: Prefetch is really nice!!'))
```

Comment:
question: since we `await` here, how do we know that the fetch isn't triggered here? I think if you remove the prefetching, this test would still pass, no?

Reply:
Yes, you are right.

I think I need to pass an alternative spy here too, to assert that the prefetch is the one calling the right function!

Reply:
Yes, you are right.

I think I need to pass an alternative spy here too, to assert that the prefetch is the one calling the right function!